codalab/lib/bundle_cli.py

```diff
@@ -1491,53 +1515,73 @@ def follow_targets(self, client, bundle_uuid, subpaths):
+            info = client.get_bundle_info(bundle_uuid)
+            if info['state'] in (State.RUNNING, State.READY, State.FAILED):
+                break
+            time.sleep(period)
+
         info = None
+        run_finished = False
         while True:
-            # Call update functions
+            if not run_finished:
+                info = client.get_bundle_info(bundle_uuid)
+                run_finished = info['state'] in (State.READY, State.FAILED)
+
+            # Read data.
             change = False
-            for i, handle in enumerate(handles):
-                if not handle:
-                    handle = handles[i] = client.open_target_handle((bundle_uuid, subpaths[i]))
-                    if not handle: continue
-                    # Go to near the end of the file (TODO: make this match up with lines)
-                    pos = max(handle.tell() - 64, 0)
-                    handle.seek(pos, 0)
-                # Read from that file
+            for i in xrange(0, len(subpaths)):
+                # If the subpath we're interested in appears, check if it's a
+                # file and if so, initialize the offset.
+                if subpath_is_file[i] is None:
+                    target_info = client.get_target_info((bundle_uuid, subpaths[i]), 0)
+                    if target_info is not None:
+                        if target_info['type'] == 'file':
+                            subpath_is_file[i] = True
+                            # Go to near the end of the file (TODO: make this match up with lines)
+                            subpath_offset[i] = max(target_info['size'] - 64, 0)
+                        else:
+                            subpath_is_file[i] = False
+
+                if not subpath_is_file[i]:
+                    continue
+
+                # Read from that file.
                 while True:
-                    result = handle.read(16384)
-                    if result == '': break
+                    READ_LENGTH = 16384
+                    result = client.read_file_section((bundle_uuid, subpaths[i]), subpath_offset[i], READ_LENGTH)
+                    if not result:
+                        break
                     change = True
+                    subpath_offset[i] += len(result)
                     self.stdout.write(result)
+                    if len(result) < READ_LENGTH:
+                        # No more to read.
+                        break
+
             self.stdout.flush()
 
-            # Update bundle info
-            info = client.get_bundle_info(bundle_uuid)
-            if info['state'] in (State.READY, State.FAILED): break
+            # The run finished and we read all the data.
+            if run_finished:
+                break
 
-            # Sleep if nothing happened
+            # If we got no data this time around, check less often.
             if not change:
-                time.sleep(period)
                 period = min(backoff*period, max_period)
 
```

Comment:
The point of the back-off was to poll the file at the rate that the file is changing. 
Anyway, I just got rid of it entirely. It's not that expensive to check for any new file contents every second.